### PR TITLE
[12.0][FIX] contract_sale_invoicing: Control singleton error on cronjob on recurring

### LINKED ISSUE #508: [12.0] [BUG]Expected singleton: contract.contract
Odoo Server Error
Traceback (most recent call last):
  File "/home/john/Developments/odoo12/odoo/odoo/models.py", line 4741, in ensure_one
    _id, = self._ids
ValueError: too many values to unpack (expected 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/john/Developments/odoo12/odoo/odoo/tools/safe_eval.py", line 350, in safe_eval
    return unsafe_eval(c, globals_dict, locals_dict)
  File "", line 1, in <module>
  File "/home/john/Developments/odoo12/extra-addons/oca/contract/contract/models/contract.py", line 483, in cron_recurring_create_invoice
    return contracts_to_invoice._recurring_create_invoice(date_ref)
  File "/home/john/Developments/odoo12/extra-addons/oca/contract/contract_sale_invoicing/models/contract.py", line 19, in _recurring_create_invoice
    if not self.invoicing_sales:
  File "/home/john/Developments/odoo12/odoo/odoo/fields.py", line 1003, in __get__
    record.ensure_one()
  File "/home/john/Developments/odoo12/odoo/odoo/models.py", line 4744, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: contract.contract(4, 7)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/john/Developments/odoo12/odoo/odoo/http.py", line 656, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/john/Developments/odoo12/odoo/odoo/http.py", line 314, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/home/john/Developments/odoo12/odoo/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/home/john/Developments/odoo12/odoo/odoo/http.py", line 698, in dispatch
    result = self._call_function(**self.params)
  File "/home/john/Developments/odoo12/odoo/odoo/http.py", line 346, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/home/john/Developments/odoo12/odoo/odoo/service/model.py", line 98, in wrapper
    return f(dbname, *args, **kwargs)
  File "/home/john/Developments/odoo12/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/home/john/Developments/odoo12/odoo/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/home/john/Developments/odoo12/odoo/odoo/http.py", line 519, in response_wrap
    response = f(*args, **kw)
  File "/home/john/Developments/odoo12/odoo/addons/web/controllers/main.py", line 966, in call_button
    action = self._call_kw(model, method, args, {})
  File "/home/john/Developments/odoo12/odoo/addons/web/controllers/main.py", line 954, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/home/john/Developments/odoo12/odoo/odoo/api.py", line 759, in call_kw
    return _call_kw_multi(method, model, args, kwargs)
  File "/home/john/Developments/odoo12/odoo/odoo/api.py", line 746, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/john/Developments/odoo12/odoo/odoo/addons/base/models/ir_cron.py", line 76, in method_direct_trigger
    self.sudo(user=cron.user_id.id).ir_actions_server_id.run()
  File "/home/john/Developments/odoo12/odoo/odoo/addons/base/models/ir_actions.py", line 553, in run
    res = func(action, eval_context=eval_context)
  File "/home/john/Developments/odoo12/odoo/addons/website/models/ir_actions.py", line 57, in run_action_code_multi
    res = super(ServerAction, self).run_action_code_multi(action, eval_context)
  File "/home/john/Developments/odoo12/odoo/odoo/addons/base/models/ir_actions.py", line 444, in run_action_code_multi
    safe_eval(action.sudo().code.strip(), eval_context, mode="exec", nocopy=True)  # nocopy allows to return 'action'
  File "/home/john/Developments/odoo12/odoo/odoo/tools/safe_eval.py", line 373, in safe_eval
    pycompat.reraise(ValueError, ValueError('%s: "%s" while evaluating\n%r' % (ustr(type(e)), ustr(e), expr)), exc_info[2])
  File "/home/john/Developments/odoo12/odoo/odoo/tools/pycompat.py", line 86, in reraise
    raise value.with_traceback(tb)
  File "/home/john/Developments/odoo12/odoo/odoo/tools/safe_eval.py", line 350, in safe_eval
    return unsafe_eval(c, globals_dict, locals_dict)
  File "", line 1, in <module>
  File "/home/john/Developments/odoo12/extra-addons/oca/contract/contract/models/contract.py", line 483, in cron_recurring_create_invoice
    return contracts_to_invoice._recurring_create_invoice(date_ref)
  File "/home/john/Developments/odoo12/extra-addons/oca/contract/contract_sale_invoicing/models/contract.py", line 19, in _recurring_create_invoice
    if not self.invoicing_sales:
  File "/home/john/Developments/odoo12/odoo/odoo/fields.py", line 1003, in __get__
    record.ensure_one()
  File "/home/john/Developments/odoo12/odoo/odoo/models.py", line 4744, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: <class 'ValueError'>: "Expected singleton: contract.contract(4, 7)" while evaluating
'model.cron_recurring_create_invoice()'

### PR DESCRIPTION
**Pull request by vinclar**
As made in other pull requests by @nikul-serpentcs like #437  #440  #442 
They commit too many improvements/features to the code, and the fix of the problem is being delayed.
This code has been made by @nikul-serpentcs  and proposed by @pedrobaeza for quicker reaction to split the fix.
If the changes are enough on this commit to solve this problem and pass the tests, I propose to solve this problem and add the new features in future commits.

### CODE REVIEW

File: contract_sale_invoicing/models/contract.py [hunk 1]
```diff
@@ -16,16 +16,19 @@ class ContractContract(models.Model):
...
-            ('invoice_status', '=', 'to invoice'),
-            ('date_order', '<=',
-             '{} 23:59:59'.format(self.recurring_next_date)),
-        ])
-        if sales:
-            invoice_ids = sales.action_invoice_create()
-            invoices |= self.env['account.invoice'].browse(invoice_ids)[:1]
+        for rec in self:
+            if not rec.invoicing_sales:
+                return invoices
```
Comment (pedrobaeza) [inline contract_sale_invoicing/models/contract.py]: You can't return in the middle of the loop

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Please check flake8 error:
```
contract_sale_invoicing/models/contract.py:34:1: W391 blank line at end of file
```
as well as the error in tests

**Comment by valero1120** [comment]
regards,
They have not yet updated the repository with the @pedrobaeza correction, what procedure is missing for them to update?

**Review by flotho — APPROVED** [review]
LGTM

**Comment by flotho** [comment]
Hi, #537 seems to be a good replacement of this one, 

what's your point of view @pedrobaeza ?

Regards

**Comment by pedrobaeza** [comment]
Yes, merging that one.

**Pull request closed without merging**